cpp/ql/src/Security/CWE/CWE-611/XXE.ql

```diff
@@ -0,0 +1,203 @@
+ */
+
+import cpp
+import semmle.code.cpp.ir.dataflow.DataFlow
+import DataFlow::PathGraph
+import semmle.code.cpp.ir.IR
+
+/**
+ * A flow state representing a possible configuration of an XML object.
+ */
+abstract class XXEFlowState extends string {
+  bindingset[this]
+  XXEFlowState() { any() } // required constructor
+}
+
+/**
+ * An `Expr` that changes the configuration of an XML object, transforming the
+ * `XXEFlowState` that flows through it.
+ */
+abstract class XXEFlowStateTranformer extends Expr {
+  /**
+   * Gets the flow state that `flowstate` is transformed into.
+   *
+   * Due to limitations of the implementation the result must always map to
+   * itself, that is, it must be that:
+   * ```
+   * transform(tranform(x)) = tranform(x)
+   * ```
+   */
+  abstract XXEFlowState transform(XXEFlowState flowstate);
+}
+
+/**
+ * The `AbstractDOMParser` class.
+ */
+class AbstractDOMParserClass extends Class {
+  AbstractDOMParserClass() { this.hasName("AbstractDOMParser") }
+}
+
+/**
+ * The `XercesDOMParser` class.
+ */
+class XercesDOMParserClass extends Class {
+  XercesDOMParserClass() { this.hasName("XercesDOMParser") }
+}
+
+/**
+ * Gets a valid flow state for `XercesDOMParser` flow.
+ *
+ * These flow states take the form `XercesDOM-A-B`, where:
+ *  - A is 1 if `setDisableDefaultEntityResolution` is `true`, 0 otherwise.
+ *  - B is 1 if `setCreateEntityReferenceNodes` is `true`, 0 otherwise.
+ */
+predicate encodeXercesDOMFlowState(string flowstate, int a, int b) {
+  flowstate = "XercesDOM-0-0" and a = 0 and b = 0
+  or
+  flowstate = "XercesDOM-0-1" and a = 0 and b = 1
+  or
+  flowstate = "XercesDOM-1-0" and a = 1 and b = 0
+  or
+  flowstate = "XercesDOM-1-1" and a = 1 and b = 1
+}
+
+/**
+ * A flow state representing the configuration of a `XercesDOMParser` object.
+ */
+class XercesDOMParserFlowState extends XXEFlowState {
+  XercesDOMParserFlowState() { encodeXercesDOMFlowState(this, _, _) }
+}
+
+/**
+ * Flow state transformer for a call to
+ * `AbstractDOMParser.setDisableDefaultEntityResolution`. Transforms the flow
+ * state through the qualifier according to the setting in the parameter.
+ */
+class DisableDefaultEntityResolutionTranformer extends XXEFlowStateTranformer {
+  Expr newValue;
+
+  DisableDefaultEntityResolutionTranformer() {
+    exists(Call call, Function f |
+      call.getTarget() = f and
+      f.getDeclaringType() instanceof AbstractDOMParserClass and
+      f.hasName("setDisableDefaultEntityResolution") and
+      this = call.getQualifier() and
+      newValue = call.getArgument(0)
+    )
+  }
+
+  final override XXEFlowState transform(XXEFlowState flowstate) {
+    exists(int a, int b |
+      encodeXercesDOMFlowState(flowstate, a, b) and
+      (
+        newValue.getValue().toInt() = 1 and // true
+        encodeXercesDOMFlowState(result, 1, b)
+        or
+        not newValue.getValue().toInt() = 1 and // false or unknown
+        encodeXercesDOMFlowState(result, 0, b)
```

Comment:
Two comments about this:
1. Collapsing the `false` and `unknown` cases into one means that we could have FPs if the value is set to a value that we can't deduce, right?

2. Would we catch more good cases by using local dataflow to find all the values that flows into `newValue`? For instance, currently, we would claim that  this is bad, I think:
```cpp
XercesDOMParser* p = ...;
p->setDisableDefaultEntityResolution(complex_condition ? true : false);
```

Reply:
> 1. Collapsing the `false` and `unknown` cases into one means that we could have FPs if the value is set to a value that we can't deduce, right?

Indeed, there's a design choice here of whether we want FPs or FNs resulting from non-constant expressions we can't determine.  Usually we would choose FNs over FPs, but in this case I feel the point of the query is to be _confident_ that the XML object is set up correctly - and that requires rejecting unknown expressions.

I may well be persuaded otherwise when we have more real world results.  We don't yet have many to judge this issue on.

> 2. Would we catch more good cases by using local dataflow to find all the values that flows into `newValue`? For instance, currently, we would claim that this is bad, I think:

Good idea.  I've added it to the TODO list on the XXE query ticket.